### DISCUSSION TITLE
Customization of integration for dynamic constraints.

### DISCUSSION THREAD

**Discussion by IlyaOrson** [infiniteopt/InfiniteOpt.jl#169] [Q&A] [answered]
I am exploring this package a bit for some optimal control usecases. From a [JuMP control example](https://jump.dev/JuMP.jl/stable/tutorials/nonlinear/rocket_control/), the integration rule for the dynamic constraints has to be hard coded in the constraint. I understand `InfiniteOpt.jl` does this step under the hood, like in [this control example](https://infiniteopt.github.io/InfiniteOpt.jl/dev/examples/Optimal%20Control/hovercraft/).

Is it possible to see or control which integration method is being used to enforce the dynamic constraints?

**Comment by pulsipher** [comment, marked as answer]
The short answer is yes, all the aspects of how the formulation is approximated can be controlled and adjusted. Moreover, there are quite a few ways to do it.

Note that by default we use direct transcription (i.e., discretization techniques) to reformulate the continuous time problem. The methodology behind this is explained in our docs [here](https://pulsipher.github.io/InfiniteOpt.jl/stable/guide/transcribe/#Transcription-Theory) and in Section 3.1 of our paper (https://arxiv.org/pdf/2106.12689.pdf).

A good place to start with getting more familiar with the syntax would be to read the "Basic Usage" section on each page of the "User Guide" in the docs. 

To get started I'll mention 3 typical areas of adjustment. Here the context will be all in the realm of optimal control as per your question.

**Specifying Support (i.e., discretization) Points**
We can specify which time points we would like to use via the `supports` keyword argument in `@infinite_parameter` or we can it have generate a certain number of uniformly spaced points via the `num_supports` keyword argument. More info is given [here](https://pulsipher.github.io/InfiniteOpt.jl/stable/guide/parameter/#Supports). These can be queried via the `supports` function.

**Specifying the Derivative Approximation Method**
Derivatives are treated as variables in InfiniteOpt such that we can support arbitrary DAEs. Thus, auxiliary equations are added to make these derivative variables well-posed. These equations are determined by the selected approximation scheme for the derivatives. We natively support finite difference schemes and orthogonal collocation over finite elements (the default is backward finite difference). The derivative approximation scheme is specified via the `derivative_method` keyword argument in the `@infinite_parameter` call. Much more info is given [here](https://pulsipher.github.io/InfiniteOpt.jl/stable/guide/derivative/).  These can be previewed via the `derivative_constraints` function (though `evaluate_derivative` must be called first).

**Specifying the Integral Evaluation Method**
We can choose which integral approximation scheme we'd like to use for a particular integral via the `eval_method` keyword argument. By default, all integral use the trapezoid rule which utilizes the time points specified in the problem (namely those given in `supports` or `num_supports`). However, a variety of other approximations are provided. Many more details are given [here](https://pulsipher.github.io/InfiniteOpt.jl/stable/guide/measure/). The approximation can be previewed via the `expand` function.

I will also mention that all the above can be extended such that we can invoke custom approximation schemes if desired.
